File: src/library/java/gg/generations/rarecandy/renderer/rendering/Bone.java
package gg.generations.rarecandy.renderer.rendering;

import gg.generations.rarecandy.assimp.AIBone;
import gg.generations.rarecandy.pokeutils.ModelNode;
import gg.generations.rarecandy.renderer.animation.Skeleton;
import org.joml.Matrix4f;

import java.util.Objects;

/**
 * We re-use this structure for multiple meshes so if you are accessing this value from outside a meshes bone array do NOT trust it.
 */
public class Bone {

    public String name;

    public Matrix4f inverseBindMatrix;
    public Matrix4f restPose;

    @Override
    public String toString() {
        return "Bone{" + "name='" + name + '\'' + '}';
    }

    @Override
    public int hashCode() {
        return Objects.hash(name);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Bone bone)) return false;

        return name.equals(bone.name);
    }

    public Bone(String name) {
        this.name = name;
        this.inverseBindMatrix = new Matrix4f();
        this.restPose = new Matrix4f();
    }
}

File: src/library/java/gg/generations/rarecandy/renderer/rendering/Bone.kt
package gg.generations.rarecandy.renderer.rendering

import org.joml.Matrix4f
import java.util.*

/**
 * We re-use this structure for multiple meshes so if you are accessing this value from outside a meshes bone array do NOT trust it.
 */
data class Bone @JvmOverloads constructor(@JvmField var name: String, @JvmField val inverseBindMatrix: Matrix4f? = Matrix4f(), @JvmField var restPose: Matrix4f? = Matrix4f()) {

    override fun toString(): String = "Bone{name='$name'}"

    override fun hashCode(): Int = name.hashCode()

    override fun equals(o: Any?): Boolean = if (this === o) true else if (o !is Bone) false else name == o.name
}
